plugins/cilium-cni/lib/deletion_queue.go

```diff
@@ -110,35 +114,54 @@ func (dc *DeletionFallbackClient) tryQueueLock() error {
 	if err != nil {
 		return fmt.Errorf("failed to acquire lock: %w", err)
 	}
+
+	dc.logger.Debug("Deletion Queue lock acquired",
+		logfields.Path, defaults.DeleteQueueLockfile,
+		logfields.Duration, time.Since(startTime))
 	dc.lockfile = lf
 	return nil
 }
 
-// EndpointDelete deletes an endpoint given by an endpoint id, either
-// by directly accessing the API or dropping in a queued-deletion file.
-// endpoint-id is a qualified endpoint reference, e.g. "container-id:XXXXXXX"
-func (dc *DeletionFallbackClient) EndpointDelete(id string) error {
-	if dc.cli != nil {
-		return dc.cli.EndpointDelete(id)
-	}
-
-	// fall-back mode
-	if dc.lockfile != nil {
-		dc.logger.Info(
-			"Queueing deletion request for endpoint",
-			logfields.EndpointID, id,
-		)
-		return dc.enqueueDeletionRequestLocked(id)
-	}
-
-	return errors.New("attempt to delete with no valid connection")
-}
-
 // EndpointDeleteMany deletes multiple endpoints based on the endpoint deletion request,
 // either by directly accessing the API or dropping in a queued-deletion file.
 func (dc *DeletionFallbackClient) EndpointDeleteMany(req *models.EndpointBatchDeleteRequest) error {
 	if dc.cli != nil {
-		return dc.cli.EndpointDeleteMany(req)
+		startTime := time.Now()
+		err := dc.cli.EndpointDeleteMany(req)
+		if err == nil {
+			dc.logger.Debug("Endpoint delete request successful",
+				logfields.Request, req,
+				logfields.Duration, time.Since(startTime),
+			)
+			return nil
+		}
+
+		status, ok := err.(runtime.ClientResponseStatus)
+		if !ok || !status.IsCode(http.StatusServiceUnavailable) {
+			// Propagate unhandled Endpoint delete errors.
+			return err
+		}
+
+		// Fallback to queuing endpoint delete if cilium Endpoint service is unavailable.
+		// Once ready, cilium-agent will process all the pending deletes in this queue.
+		dc.logger.Warn(
+			"Received ServiceUnavailable response from server for batch Endpoint Delete request",
+			logfields.Request, req,
+		)
+
+		// If we don't already have lock on the local endpoint deletion queue, try acquire it.
+		if dc.lockfile == nil {
+			dc.logger.Debug("Deletion Queue lock not already held, reprocessing delete request with locked queue")
+			if err = dc.tryQueueLock(); err != nil || dc.lockfile == nil {
+				// LockFile nil check to make sure we don't end up in an inifinite loop.
+				return fmt.Errorf("failed to acquire deletion queue lock: %w", err)
+			}
+
+			// Once we have the lock reprocess the Endpoint delete request. This is to make sure
+			// we don't race with an already proccessed cilium-agent endpoint deletion queue while
+			// we waited to acquire the lock.
+			return dc.EndpointDeleteMany(req)
```

Comment:
Can you explain the race further here? In the logic I trace:
1. A connection was established to cilium-agent
2. cilium-agent is emitting 503s
3. cilium-agent finishes processing its existing deletion queue and unlocks the lockfile
4. This goroutine grabs the lock on the lockfile
5. Now there is a recursive call

This seems like it should just `dc.cli.EndpointDeleteMany(req)` instead?